exp/bufferpool/pool.go

```diff
@@ -1,52 +1,135 @@
 var Default Pool
 
-// A pool of buffers, in variable sizes.
+// Pool maintains a list of BucketCount buckets that contain buffers
+// of exponentially-increasing capacity, 1 << 0 to 1 << BucketCount.
+//
+// The MinAlloc field specifies the minimum capacity of new buffers
+// allocated by Pool, which improves reuse of small buffers. For the
+// avoidance of doubt:  calls to Get() with size < MinAlloc return a
+// buffer of len(buf) = size and cap(buf) >= MinAlloc. MinAlloc MUST
+// NOT exceed 1 << BucketCount, or method calls to Pool will panic.
+//
+// The zero-value Pool is ready to use, defaulting to BucketCount=20
+// and MinAlloc=1024 (max size = ~1MiB).  Most applications will not
+// benefit from tuning these parameters.
+//
+// As a general rule, increasing MinSize reduces GC latency at the
+// expense of increased memory usage.  Increasing BucketCount can
+// reduce GC latency in applications that frequently allocate large
+// buffers.
 type Pool struct {
-	buckets [nBuckets]sync.Pool
+	once                  sync.Once
+	MinAlloc, BucketCount int
+	buckets               bucketSlice
 }
 
-// Get a buffer of the given length. Its capacity may be larger than the
-// requested size.
+// Get a buffer of len(buf) == size and cap >= size.
 func (p *Pool) Get(size int) []byte {
-	for i := 0; i < nBuckets; i++ {
-		capacity := 1 << i
-		if capacity >= size {
-			var ret []byte
-			item := p.buckets[i].Get()
-			if item == nil {
-				ret = make([]byte, capacity)
-			} else {
-				ret = item.([]byte)
-			}
-			ret = ret[:size]
-			return ret
-		}
+	p.init()
+
+	if buf := p.buckets.Get(size); buf != nil {
+		return buf[:size]
 	}
+
 	return make([]byte, size)
 }
 
-// Return a buffer to the pool. Zeros the slice (but not the full backing array)
-// before making it available for future use.
+// Put returns the buffer to the pool.  The first len(buf) bytes
+// of the buffer are zeroed.
 func (p *Pool) Put(buf []byte) {
-	for i := 0; i < len(buf); i++ {
+	p.init()
+
+	for i := range buf {
 		buf[i] = 0
 	}
 
-	capacity := cap(buf)
-	for i := 0; i < nBuckets; i++ {
-		if (1 << i) == capacity {
-			p.buckets[i].Put(buf[:capacity])
-			return
+	p.buckets.Put(buf[:cap(buf)])
+}
+
+func (p *Pool) init() {
+	p.once.Do(func() {
+		if p.MinAlloc <= 0 {
+			p.MinAlloc = defaultMinSize
+		}
+
+		if p.BucketCount <= 0 {
+			p.BucketCount = defaultBucketCount
+		}
+
+		if p.MinAlloc > (1 << p.BucketCount) {
+			panic("MinSize greater than largest bucket")
+		}
+
+		p.buckets = make(bucketSlice, p.BucketCount)
+		for i := range p.buckets {
+			p.buckets[i] = newBucket(i)
+		}
+
+		// Find the index of the bucket responsible for MinAlloc.
+		var idx int
+		for idx = range p.buckets {
+			if 1<<idx >= p.MinAlloc {
+				break
+			}
+		}
+
+		// Set the 'New' function for all buckets lower than 'MinSize'
+		// to the MinSize bucket's 'New' function.  This ensures that
+		// we always allocate a capacity of *at least* MinSize, while
+		// still re-using small buffers.
```

Comment:
That would result in undersized buffers being returned to `p.buckets[idx]`.

Reply:
I've reworked this a bit to make it more readable.

Reply:
But if we're returning oversized buffers anyway, shouldn't it be fine?

Reply:
Per discussion on matrix, for posterity: this is so that it's possible to return buffers not allocated with the Pool in the first place, so we'll keep it as-is.